### PR TITLE
Implement the etag functionality in the `readUrl` method of `FetchUrlReader`

### PR DESCRIPTION
**Pull request by dhenneke**
## Hey, I just made a Pull Request!

Follow up of https://github.com/backstage/backstage/pull/6305 to add the caching logic to the `FetchUrlReader`. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by OrkoHunter — APPROVED** [review]
I always wonder what FetchUrlReader does at the first glance. 😄 
>  A UrlReader that does a plain fetch of the URL.

Looks good to me.

File: packages/backend-common/src/reading/FetchUrlReader.ts [hunk 1]
```diff
@@ -57,15 +57,34 @@ export class FetchUrlReader implements UrlReader {
...
+      response = await fetch(url, {
+        headers: {
+          ...(options?.etag && { 'If-None-Match': options.etag }),
+        },
+      });
     } catch (e) {
       throw new Error(`Unable to read ${url}, ${e}`);
     }
 
+    if (response.status === 304) {
```
Comment (benjdlambert) [inline packages/backend-common/src/reading/FetchUrlReader.ts:79]: It's unfortunate that we have to throw an error when things are happy path, but I guess it's really the only way that we can implement this without having the previous response to hand.

File: packages/backend-common/src/reading/FetchUrlReader.ts [hunk 2]
```diff
@@ -57,15 +57,34 @@ export class FetchUrlReader implements UrlReader {
...
+      throw new NotModifiedError();
+    }
+
     if (response.ok) {
-      return Buffer.from(await response.text());
+      return {
+        buffer: async () => Buffer.from(await response.text()),
+        etag: response.headers.get('ETag') ?? undefined,
```
Comment (benjdlambert) [inline packages/backend-common/src/reading/FetchUrlReader.ts:86]: Can we lowercase these headers and it all still work fine or are they parsed and cap'd by node?

Comment (dhenneke) [reply packages/backend-common/src/reading/FetchUrlReader.ts:86]: We could. Headers are case-insensitive by the RFC and they always converted to lowercase in node internally. I started with lowercase names but I used the pattern that was established in the [GithubUrlReader](https://github.com/backstage/backstage/blob/2bdf2dd143c3a0de4cff03b6b4a549c52401e544/packages/backend-common/src/reading/GithubUrlReader.ts#L114). Not sure why @Rugvip decided to do that. At least `ETag` is the official name in the ([RFC7232#2.3](https://datatracker.ietf.org/doc/html/rfc7232#section-2.3)).

Comment (benjdlambert) [reply packages/backend-common/src/reading/FetchUrlReader.ts:86]: meh. ok. if there's prior art for it let's keep it this way. 😅

Comment (Rugvip) [reply packages/backend-common/src/reading/FetchUrlReader.ts:86]: I find that header literals are generally title cased unless accessed via a case sensitive API, which no header API should be 😅. A bit nicer to read and often how they are sent over the wire as well

**Pull request merged**